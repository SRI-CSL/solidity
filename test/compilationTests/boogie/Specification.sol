pragma solidity ^0.4.23;

/**
 * @notice invariant x == y
 * @notice some random comment
 */
contract SomeContract {
    uint x;
    uint y;

    constructor() public {
        x = y = 0;
    }

    function add(uint amount) public {
        x += amount;
        y += amount;
    }

    function add_incorrect(uint amount) public {
        x += amount;
        y -= amount; // ERROR
    }

    function add_incorrect_2(uint amount) public {
        x += amount;
        add(amount); // ERROR, invariant does not hold when calling function
        y += amount;
    }
}

/**
 * @notice invariant x == y
 */
contract DefaultConstructor {
    // The generated default constructor (setting the initial values) should fail
    uint x = 0;
    uint y = 1;

    function add(uint amount) public {
        x += amount;
        y += amount;
    }
}