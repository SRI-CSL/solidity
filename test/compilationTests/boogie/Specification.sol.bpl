// Global declarations and definitions related to the address type
type address_t;
var __balance: [address_t]int;
procedure __transfer(__this: address_t, __msg_sender: address_t, __msg_value: int, amount: int)
  requires (__balance[__msg_sender] >= amount);
  ensures ((__this != __msg_sender) ==> (__balance[__msg_sender] == (old(__balance[__msg_sender]) - amount)));
  ensures ((__this != __msg_sender) ==> (__balance[__this] == (old(__balance[__this]) + amount)));
  ensures ((__this == __msg_sender) ==> (__balance[__msg_sender] == old(__balance[__msg_sender])));
  ensures ((__this == __msg_sender) ==> (__balance[__this] == old(__balance[__this])));
{
  __balance := __balance[__this := (__balance[__this] + amount)];
  __balance := __balance[__msg_sender := (__balance[__msg_sender] - amount)];
  // TODO: call fallback, exception handling
}

procedure __call(__this: address_t, __msg_sender: address_t, __msg_value: int)
  returns (__result: bool)
{
  // TODO: model something nondeterministic here
}

procedure __send(__this: address_t, __msg_sender: address_t, __msg_value: int, amount: int)
  returns (__result: bool)
  requires (__balance[__msg_sender] >= amount);
  ensures ((__result && (__this != __msg_sender)) ==> (__balance[__msg_sender] == (old(__balance[__msg_sender]) - amount)));
  ensures ((__result && (__this != __msg_sender)) ==> (__balance[__this] == (old(__balance[__this]) + amount)));
  ensures ((!(__result) || (__this == __msg_sender)) ==> (__balance[__msg_sender] == old(__balance[__msg_sender])));
  ensures ((!(__result) || (__this == __msg_sender)) ==> (__balance[__this] == old(__balance[__this])));
{
  // TODO: call fallback
  if (*) {
  __balance := __balance[__this := (__balance[__this] + amount)];
  __balance := __balance[__msg_sender := (__balance[__msg_sender] - amount)];
  __result := true;
  }
  else {
  __result := false;
  }

}

type string_t;
// 
// ------- Source: Specification.sol -------
// Pragma: solidity^0.4.23
// 
// ------- Contract: SomeContract -------
// 
// State variable: x : uint256
var x#3: [address_t]int;
// 
// State variable: y : uint256
var y#5: [address_t]int;
// 
// Function:  : function ()
procedure __constructor#15(__this: address_t, __msg_sender: address_t, __msg_value: int)
{
  y#5 := y#5[__this := 0];
  x#3 := x#3[__this := y#5[__this]];
  $return0:
}

// 
// Function: add : function (uint256)
procedure add#29(__this: address_t, __msg_sender: address_t, __msg_value: int, amount#17: int)
{
  x#3 := x#3[__this := (x#3[__this] + amount#17)];
  y#5 := y#5[__this := (y#5[__this] + amount#17)];
  $return1:
}

