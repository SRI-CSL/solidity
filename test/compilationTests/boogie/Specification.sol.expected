^Boogie program verifier finished with 6 verified, 3 errors$